### PR TITLE
Invalidate coupon cache after hold/usage data is modified in datastore via SQL

### LINKED ISSUE #41506: Coupon usage restriction not properly respected for guest orders under HPOS
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

Coupon usage restriction not properly respected for guest orders under HPOS. 

### Expected behavior

When a guest tries to use a coupon that can only be used once per user in two different orders that are made with no account, but with the same email address, WooCommerce should reject the second coupon usage. 

### Actual behavior

This works as expected under legacy storage. Under HPOS the order is accepted with the coupon. 

### Steps to reproduce

0. Have a store set up with products and a payment method, ensure order storage is set to HPOS. 
1. Create a coupon with 10% off, individual use only, exclude sales items, usage limit per user set to 1.
3. In an incognito window, order something as a guest and with the coupon applied. Note that the order goes through as expected. 
4. In a new incognito window, order something as a guest, with the coupon applied, and with the same email address from step 3. Note that the order goes through, even though it is not expected to. 

### WordPress Environment

<details>

### WordPress Environment ###

WordPress address (URL): https://**********
Site address (URL): https://**********
WC Version: 8.2.2
REST API Version: ✔ 8.2.2
WC Blocks Version: ✔ 11.1.3
Action Scheduler Version: ✔ 3.6.3
Log Directory Writable: ✔
WP Version: 6.4.1
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.57 (Unix) OpenSSL/1.0.2g
PHP Version: 8.0.30
PHP Post Max Size: 1 GB
PHP Time Limit: 30
PHP Max Input Vars: 5000
cURL Version: 7.47.0
OpenSSL/1.0.2g

SUHOSIN Installed: –
MySQL Version: 5.7.33-0ubuntu0.16.04.1-log
Max Upload Size: 512 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 8.2.2
WC Database Prefix: wp_
Total Database Size: 5.49MB
Database Data Size: 3.68MB
Database Index Size: 1.81MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_jetpack_sync_queue: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 2.48MB + Index: 0.08MB + Engine InnoDB
wp_postmeta: Data: 0.11MB + Index: 0.06MB + Engine InnoDB
wp_posts: Data: 0.06MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_orders: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_wc_orders_meta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_addresses: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_operational_data: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 24
page: 7
post: 2
product: 18
product_variation: 7
shop_coupon: 1
shop_order_placehold: 2
wp_navigation: 1
wp_template: 2

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (3) ###

Companion Plugin: by Osk – 1.30
Jetpack: by Automattic – 12.8.1
WooCommerce: by Automattic – 8.2.2

### Inactive Plugins (2) ###

Akismet Anti-spam: Spam Protection: by Automattic - Anti-spam Team – 5.3
Hello Dolly: by Matt Mullenweg – 1.7.2

### Settings ###

API Enabled: –
Force SSL: –
Currency: EUR (€)
Currency Position: right_space
Thousand Separator: .
Decimal Separator: ,
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔
HPOS feature screen enabled: ✔
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: –

### WC Pages ###

Shop base: #7 - /shop/
Cart: #8 - /cart/
Checkout: #9 - /checkout/
My account: #10 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty-Four
Version: 1.0
Author URL: https://wordpress.org
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ❌ Not declared

### Templates ###

Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
product-variation-management
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: customize-store
minified-js
new-product-management-experience
settings
async-product-editor-category-field

Daily Cron: ✔ Next scheduled: 2023-11-17 12:41:51 +00:00
Options: ✔
Notes: 61
Onboarding: completed

### Action Scheduler ###

Complete: 32
Oldest: 2023-11-16 12:42:57 +0000
Newest: 2023-11-16 12:48:46 +0000

Failed: 1
Oldest: 2023-11-16 12:45:59 +0000
Newest: 2023-11-16 12:45:59 +0000

Pending: 1
Oldest: 2023-11-17 12:42:57 +0000
Newest: 2023-11-17 12:42:57 +0000

### Status report information ###

Generated at: 2023-11-16 12:51:19 +00:00

</details>

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woo.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by jorgeatorres**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

The coupons datastore uses a mix of calls to post meta functions and also direct SQL queries to performantly record usage of coupons as well as coupons held during checkout, etc. While this works for performance, it does mean that instantiating a given coupon at a later stage might not reflect the correct information in the database, because either the internal `WC_Data` cache does not pick up changes made outside of CRUD, or WP's postmeta those made via SQL queries.

Some of the methods in the datastore did attempt to do some cache clearing, but that was insufficient.

This PR makes sure we're a bit more aggressive when invalidating coupon-related cache, at least until the time comes that we properly implement these methods as CRUD methods.

<details><summary>Why did this result in coupon usage not being recorded with HPOS?</summary>
<p>

A curious one you are, young Padawan.

When coupon usage is updated (like [here](https://github.com/woocommerce/woocommerce/blob/c768ca0dc680cf64c596c76738251e39fb7f6c91/plugins/woocommerce/includes/data-stores/class-wc-coupon-data-store-cpt.php#L334C1-L342)) we some times use a direct query. In this case, the query actually _renames_ the meta key of a row of metadata to `_used_by`, but given caches were not cleared, our CRUD layer didn't pick up this change.

Instead, instances of said `WC_Coupon` would still think that row used the prior `meta_key` name. That's not a problem most of the time, except the HPOS datastore, contrary to the CPT one, [uses CRUD methods to release coupons held during checkout](https://github.com/woocommerce/woocommerce/blob/c768ca0dc680cf64c596c76738251e39fb7f6c91/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php#L2651-L2669).
Metadata in `WC_Data` is deleted [by referencing the `meta_id`](https://github.com/woocommerce/woocommerce/blob/c768ca0dc680cf64c596c76738251e39fb7f6c91/plugins/woocommerce/includes/abstracts/abstract-wc-data.php#L685C4-L686) value of any given piece of metadata, and since the SQL query doesn't change this, HPOS was effectively removing the data that was no longer temporary and we actually wanted to keep.

In this case, the record of the coupon being used.
</p>
</details> 

Closes #41506.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

0. Make sure HPOS is active in WC > Settings > Advanced > Features.
1. Go to Marketing > Coupons and create a coupon with any details you like. Make sure it's limited to 1 use per user.
2. In an incognito window, order something as a guest and with the coupon applied.
3. Repeat step 2 and confirm that the order doesn't go through this time.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

**Comment by jorgeatorres** [comment]
Just noting that it doesn't really matter if the coupon is limited in any way. The important bit is to confirm that `_used_by` for the coupon is properly updated after every use, which could be verified with

```wp post meta list COUPON_ID --keys=_used_by```

The user ID (if logged-in) should be recorded. If not, then their e-mail.

File: plugins/woocommerce/includes/data-stores/class-wc-coupon-data-store-cpt.php [hunk 1]
```diff
@@ -378,6 +380,8 @@ public function decrease_usage_count( &$coupon, $used_by = '' ) {
 			}
 		}
 
+		$this->refresh_coupon_data( $coupon );
```
Comment (lsinger) [inline plugins/woocommerce/includes/data-stores/class-wc-coupon-data-store-cpt.php]: I'm not sure if I understand the issue fully, but while `decrease_usage_count` does contain raw SQL, that's only a `SELECT`. The actual changes (and therefore the cache invalidation) are happening via `update_usage_count_meta`, `delete_metadata_by_mid`, and `$coupon->set_used_by`. 

I see three lines in which we change data in this function:

- `$this->update_usage_count_meta` (this already calls `refresh_coupon_data` itself)
- `delete_metadata_by_mid` (calls `wp_cache_delete`)
- `$coupon->set_used_by` (calls `set_prop`, which does not actually change data in the database -- not super clear to me when and where this actually gets written to the database?)

Do we really need to `refresh_coupon_data` here? 🤔 Or do we maybe only need it inside of the `if ( $meta_id )` block?

File: plugins/woocommerce/tests/php/src/Internal/DataStores/Orders/OrdersTableDataStoreTests.php [hunk 2]
```diff
@@ -3198,4 +3198,45 @@ public function test_order_updated_webhook_delivered_once() {
...
+
+		WC()->cart->add_to_cart( $product->get_id(), 1 );
+		WC()->cart->add_discount( $coupon->get_code() );
+
+		$this->assertEquals( 0, $coupon->get_data_store()->get_usage_by_email( $coupon, 'user@woo.test' ) );
+
+		$order_id = WC()->checkout->create_order(
+			array(
+				'billing_email'  => 'user@woo.test',
+				'payment_method' => 'dummy'
```
Comment (lsinger) [inline plugins/woocommerce/tests/php/src/Internal/DataStores/Orders/OrdersTableDataStoreTests.php]: The linter [really, really wants](https://github.com/woocommerce/woocommerce/actions/runs/6903967483/job/18783779604?pr=41538) this to end in a comma:

```suggestion
				'payment_method' => 'dummy',
```

**Review by lsinger — APPROVED** [review]
Great and super swift work @jorgeatorres, kudos to you! And thank you for the great explanation as well. ☺️ 

I left a suggestion to appease the linter and a question where I'm not sure whether the call to invalidate the cache is really needed. 

Feel free to merge once CI is green (with the linter appeased) and when you feel the call from my question should indeed be there. 

*Edit*: I forgot to note that indeed the issue is fixed when going through those steps on this branch. 👌

**Review by lsinger — APPROVED** [review]
LGTM! 🚢

**Comment by awebguy99** [comment]
I just wanted to say thanks for the super quick work on this. My clients will certainly appreciate this.

**Comment by kevin25** [comment]
I got a lot of declined payment via Stripe when I enabled checkout by Woo Blocks. Is it related to this as well?

**Comment by jorgeatorres** [comment]
Hi @kevin25! I don't think it's related as this fixes a bug in the legacy checkout, not the Blocks one. It might be a different thing altogether, but if you feel you're encountering a bug specific to the Blocks checkout, please do report it in the Blocks repo https://github.com/woocommerce/woocommerce-blocks.

**Pull request merged**